File: src/main/java/com/viartemev/requestmapper/annotations/MappingAnnotation.java
package com.viartemev.requestmapper.annotations;

import com.intellij.openapi.project.Project;
import com.intellij.psi.PsiAnnotation;
import com.intellij.psi.PsiElement;
import com.viartemev.requestmapper.RequestMappingItem;
import com.viartemev.requestmapper.annotations.spring.DeleteMapping;
import com.viartemev.requestmapper.annotations.spring.GetMapping;
import com.viartemev.requestmapper.annotations.spring.PostMapping;
import com.viartemev.requestmapper.annotations.spring.PutMapping;
import com.viartemev.requestmapper.annotations.spring.RequestMapping;
import org.apache.commons.lang.StringUtils;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.List;
import java.util.Objects;

public interface MappingAnnotation {
    List<RequestMappingItem> values();

    @NotNull
    static MappingAnnotation mappingAnnotation(Project project,
                                               @Nullable String annotationName,
                                               PsiAnnotation psiAnnotation,
                                               PsiElement psiElement) {
        if (StringUtils.isEmpty(annotationName)) {
            return UnknownAnnotation.getInstance();
        }
        if (Objects.equals(annotationName, RequestMapping.class.getSimpleName())) {
            return new RequestMapping(psiAnnotation, psiElement, project);
        } else if (Objects.equals(annotationName, GetMapping.class.getSimpleName())) {
            return new GetMapping(psiAnnotation, psiElement, project);
        } else if (Objects.equals(annotationName, PostMapping.class.getSimpleName())) {
            return new PostMapping(psiAnnotation, psiElement, project);
        } else if (Objects.equals(annotationName, PutMapping.class.getSimpleName())) {
            return new PutMapping(psiAnnotation, psiElement, project);
        } else if (Objects.equals(annotationName, DeleteMapping.class.getSimpleName())) {
            return new DeleteMapping(psiAnnotation, psiElement, project);
        }
        return UnknownAnnotation.getInstance();
    }

}

File: src/main/kotlin/com/viartemev/requestmapper/annotations/MappingAnnotation.kt
package com.viartemev.requestmapper.annotations

import com.intellij.psi.PsiAnnotation
import com.viartemev.requestmapper.RequestMappingItem
import com.viartemev.requestmapper.annotations.jaxrs.Delete
import com.viartemev.requestmapper.annotations.jaxrs.Get
import com.viartemev.requestmapper.annotations.jaxrs.Head
import com.viartemev.requestmapper.annotations.jaxrs.Options
import com.viartemev.requestmapper.annotations.jaxrs.Patch
import com.viartemev.requestmapper.annotations.jaxrs.Post
import com.viartemev.requestmapper.annotations.jaxrs.Put
import com.viartemev.requestmapper.annotations.spring.DeleteMapping
import com.viartemev.requestmapper.annotations.spring.GetMapping
import com.viartemev.requestmapper.annotations.spring.PatchMapping
import com.viartemev.requestmapper.annotations.spring.PostMapping
import com.viartemev.requestmapper.annotations.spring.PutMapping
import com.viartemev.requestmapper.annotations.spring.RequestMapping

interface MappingAnnotation {

    fun values(): List<RequestMappingItem>

    companion object {
        val supportedAnnotations = listOf(
            RequestMapping::class.java.simpleName,
            GetMapping::class.java.simpleName,
            PostMapping::class.java.simpleName,
            PutMapping::class.java.simpleName,
            PatchMapping::class.java.simpleName,
            DeleteMapping::class.java.simpleName,
            Get::class.java.simpleName,
            Put::class.java.simpleName,
            Post::class.java.simpleName,
            Options::class.java.simpleName,
            Head::class.java.simpleName,
            Delete::class.java.simpleName,
            Patch::class.java.simpleName
        )

        fun mappingAnnotation(annotationName: String, psiAnnotation: PsiAnnotation): MappingAnnotation {
            return when (annotationName) {
                RequestMapping::class.java.simpleName -> RequestMapping(psiAnnotation)
                GetMapping::class.java.simpleName -> GetMapping(psiAnnotation)
                PostMapping::class.java.simpleName -> PostMapping(psiAnnotation)
                PutMapping::class.java.simpleName -> PutMapping(psiAnnotation)
                PatchMapping::class.java.simpleName -> PatchMapping(psiAnnotation)
                DeleteMapping::class.java.simpleName -> DeleteMapping(psiAnnotation)
                Get::class.java.simpleName -> Get(psiAnnotation)
                Put::class.java.simpleName -> Put(psiAnnotation)
                Post::class.java.simpleName -> Post(psiAnnotation)
                Options::class.java.simpleName -> Options(psiAnnotation)
                Head::class.java.simpleName -> Head(psiAnnotation)
                Delete::class.java.simpleName -> Delete(psiAnnotation)
                Patch::class.java.simpleName -> Patch(psiAnnotation)
                else -> UnknownAnnotation
            }
        }
    }
}
